Documentation/gen-api.md

```diff
@@ -0,0 +1,46 @@
+
+Using GenAPI allows us to strip out internals, removing the dangling references from our reference assemblies.
+
+## [GenApi.props](/eng/WpfArcadeSdk/tools/GenApi.props)
+Contains various properties related to GenAPI runs and configurations.
+* `GenAPIEnabledProjects`
+  * The set of projects to run GenAPI on
+* `GlobalApiExclusionsFile`
+  * A file that specifies API surface area to exclude from code generation (see [GlobalApiExclusions.txt](/eng/WpfArcadeSdk/tools/GenApi/GlobalApiExclusions.txt))
+* `GlobalAttrExclusionsFile`
+  * A file that specifies Attributes to exclude from code generation (see [GlobalAttrExclusions.txt](/eng/WpfArcadeSdk/tools/GenApi/GlobalAttrExclusions.txt))
+* `GenAPIAdditionalParameters`
+  * Parameters to GenAPI built up from local configuration
+* _GenerateReferenceAssemblySource
+  * A private parameter used to enable GenAPI targets
+## [GenApi.targets](/eng/WpfArcadeSdk/tools/GenApi.targets)
+Contains targets and properties related to GenAPI runs
+* `GenAPITargetDir`
+  * The directory where GenAPI will generate code
+* `GenAPITargetPath`
+  * The full path to the file GenAPI will generate
+* `EnsureGenAPITargetDirectory`
+  * Creates the directory specified by `GenAPITargetDir` if it does not exist
+## Using GenAPI in WPF
+GenAPI is run only on-demand.  In the event that a change to a runtime assembly creates new public surface area, a developer will see an [ApiCompat](api-compat.md) error between the reference assembly and the runtime assembly.  In order to address this, the developer must run GenAPI to generate new reference assembly code.
+### Running GenAPI
+GenAPI can be run by setting the following MSBuild property while building.
+```
+/p:GenerateReferenceAssemblySource=true
+```
+When a build is run with that property enabled, GenAPI will read the runtime assembly and generate a new `{AssemblyName}.cs` file under the ref directory in the assembly's source tree.
+
+This new file will contain the newly created surface area and will need to be checked in along with the runtime assembly change.  The next build without `GenerateReferenceAssemblySource` enabled will no longer display an ApiCompat error as the surface area will now match the baseline.
+### Issues with GenAPI
+Often, GenAPI will generate code output that will contain code that is either private, internal, or creates build errors.  For this reason a developer usually cannot just use the output of GenAPI directly.  Instead, the developer should do the following:
+* Build with GenAPI enabled
+* Diff the output file against the previous version
+* Extract just the new surface area (and related code) from the generated code
+* Revert the generated file
+* Add back the new surface area to the reference assembly code
+* Ensure that nothing in the new surface area is private or internal unless requried by XAML compilation or other reference assemblies
```

Comment:
It doesn't sound like a good idea to have internal/private APIs required by the XAML compiler. I understand that IAddChildInternal exists today, but is this something we want to put in as being an ok thing to do in the future?

Reply:
This is more if you make a change to an existing class that uses something like that, the entire class will get re-generated, sans the IAddChildInternal.  So you need to ensure that you're not messing with that piece.

Reply:
We can have a discussion about fixing the docs up later, at this point I don't want to spin another build just due to doc updates.  Helix queues are killing me today.